src/reaction_network.jl

```diff
@@ -203,144 +257,196 @@ struct ReactionStruct
-    end
-    for prod in rxstruct.products
-        push!(reaction_func.args[4].args, prod.reactant)
-        push!(reaction_func.args[6].args, prod.stoichiometry)
-    end
-    reaction_func
-end
-
-function get_pexprs(psyms)
-    pexprs = isempty(psyms) ? :() : :(@parameters)
-    if !isempty(psyms)
-        foreach(psym -> push!(pexprs.args, psym), psyms)
-    end
-    pexprs
-end
+### Functions rephrasing the macro input as a ReactionSystem structure. ###
 
-function get_sexprs(ssyms)
-    sexprs = :(@variables t)
-    foreach(s -> (s isa Symbol) && push!(sexprs.args, Expr(:call, s, :t)), ssyms)
-    sexprs
+function forbidden_symbol_check(v)
+    !isempty(intersect(forbidden_symbols, v)) &&
+    error("The following symbol(s) are used as species or parameters: " *
+          ((map(s -> "'" * string(s) * "', ", intersect(forbidden_symbols, v))...)) *
+          "this is not permited.")
+    nothing
 end
 
-# Takes the reactions, and rephrases it as a "ReactionSystem" call, as designated by the ModelingToolkit IR.
-function make_reaction_system(ex::Expr, parameters; name = :(gensym(:ReactionSystem)))
+# Function for creating a ReactionSystem structure (used by the @reaction_network macro).
+function make_reaction_system(ex::Expr; name = :(gensym(:ReactionSystem)))
 
-    # handle interpolation of variables
+    # Handle interpolation of variables
     ex = esc_dollars!(ex)
 
-    # parse DSL lines
-    reactions = get_reactions(ex)
-    reactants = get_reactants(reactions)
-    allspecies = union(reactants, get_rx_species(reactions, parameters))
-    !isempty(intersect(forbidden_symbols, union(allspecies, parameters))) &&
-        error("The following symbol(s) are used as species or parameters: " *
-              ((map(s -> "'" * string(s) * "', ",
-                    intersect(forbidden_symbols, union(species, parameters)))...)) *
-              "this is not permited.")
-
-    pexprs = get_pexprs(parameters)    # parameters
-    sexprs = get_sexprs(allspecies)    # species
-
-    # ReactionSystem
-    rxexprs = :($(make_ReactionSystem_internal)([], t, nothing, []; name = $(name)))
+    # Read lines with reactions and options.
+    reaction_lines = Expr[x for x in ex.args if x.head == :tuple]
+    option_lines = Expr[x for x in ex.args if x.head == :macrocall]
+
+    # Get macro options.
+    options = Dict(map(arg -> Symbol(String(arg.args[1])[2:end]) => arg,
+                       option_lines))
+
+    # Parses reactions, species, and parameters.
+    reactions = get_reactions(reaction_lines)
+    species_declared = haskey(options, :species) ?
+                       extract_syms(options[:species], :species) : Union{Symbol,Expr}[]
+    parameters_declared = haskey(options, :parameters) ?
+                          extract_syms(options[:parameters], :parameters) :
+                          Union{Symbol,Expr}[]
+    declared_syms = Set(Iterators.flatten((parameters_declared, species_declared)))
```

Comment:
Made this a `Set` since we are repeatedly checking if something is in it.

Reply:
makes sense